### PR TITLE
chore(search): Rax TreeMap

### PR DESCRIPTION
**Pull request by dranikpg**
Self contained RaxTreeMap that hides rax functions + allocations

### CODE REVIEW

File: src/core/search/rax_tree.h [hunk 1]
```diff
@@ -0,0 +1,145 @@
...
+    void* ptr = mr_->allocate(sizeof(V), alignof(V));
+    V* data = new (ptr) V(std::forward<Args>(args)...);
+
+    V* old = nullptr;
+    raxInsert(tree_, to_key_ptr(key), key.size(), data, reinterpret_cast<void**>(&old));
+    assert(old == nullptr);
+
+    auto it = std::make_optional(std::make_pair(key, data));
+    return std::make_pair(FindIterator{it}, true);
+  }
```
Comment (dranikpg) [inline src/core/search/rax_tree.h:126]: It's so ugly 😭

**Comment by dranikpg** [comment]
I will replace the built-in absl::flat_hash_map with this data structure and we'll have prefix searches automatically, for suffix/infix searches we'll build a suffix tree on top of this struct

**Comment by dranikpg** [comment]
Forgot to mention, properly using polymorphic allocators is a pain because we use redis/zmalloc internally 😕 Not sure what to do about it 🤷🏻‍♂️

File: src/core/search/rax_tree.h [hunk 2]
```diff
@@ -0,0 +1,149 @@
...
+      return *this;
+    }
+
+    std::pair<std::string_view, V&> operator*() const {
+      return {std::string_view{reinterpret_cast<const char*>(it->key), it->key_len},
+              *reinterpret_cast<V*>(it->data)};
+    }
+
+   private:
+    std::optional<raxIterator> it;
```
Comment (romange) [inline src/core/search/rax_tree.h:67]: `it_`

**Comment by dranikpg** [comment]
Ah, I forgot our PMR_NS namespace 😅 Will fix it

File: src/core/search/rax_tree.h (hunk 2 above)
Comment (romange) [reply src/core/search/rax_tree.h:67]: nit: raxIterator can easily support the "uninitialized" state without changing rax code, so optional here is redundant.

File: src/core/search/rax_tree.h [hunk 3]
```diff
@@ -0,0 +1,149 @@
...
+// Allocates all objects on heap (with custom memory resource) as rax tree operates fully on
+// pointers.
+template <typename V> struct RaxTreeMap {
+  struct FindIterator;
+
+  // Simple seeking iterator
+  struct SeekIterator {
+    friend struct FindIterator;
+
+    SeekIterator() {
```
Comment (romange) [inline src/core/search/rax_tree.h:27]: can be `raxStart(&it, NULL)` that sets `RAX_ITER_EOF` flag to the iterator, see below

File: src/core/search/rax_tree.h [hunk 4]
```diff
@@ -0,0 +1,149 @@
...
+    std::pair<std::string_view, V&> operator*() const {
+      return {std::string_view{reinterpret_cast<const char*>(it->key), it->key_len},
+              *reinterpret_cast<V*>(it->data)};
+    }
+
+   private:
+    std::optional<raxIterator> it;
+  };
+
+  struct FindIterator : public std::optional<std::pair<std::string_view, V&>> {
```
Comment (romange) [inline src/core/search/rax_tree.h:70]: maybe add a comment that we use optional here to get "->" operators for free.

File: src/core/search/rax_tree_test.cc [hunk 5]
```diff
@@ -0,0 +1,97 @@
...
+
+  size_t i = 0;
+  for (auto [key, value] : map) {
+    EXPECT_EQ(elements[i].first, key);
+    EXPECT_EQ(elements[i].second, value);
+    i++;
+  }
+}
+
+TEST_F(RaxTreeTest, LowerBound) {
```
Comment (romange) [inline src/core/search/rax_tree_test.cc:55]: check how it works with empty string_views, strings. Sometimes redis code has problems with zero length slices that have nullptr pointers

**Review by romange — COMMENTED** [review]
LGTM

File: src/core/search/rax_tree_test.cc [hunk 6]
```diff
@@ -74,6 +75,12 @@ TEST_F(RaxTreeTest, LowerBound) {
 
   EXPECT_TRUE(it1 == map.end());
   EXPECT_TRUE(it2 == keys.end());
+
+  // Test lower bound empty string
+  vector<string> keys2;
+  for (auto it = map.lower_bound(""); it != map.end(); ++it)
```
Comment (romange) [inline src/core/search/rax_tree_test.cc:81]: it's not enough. 
`""` actually is non-null string_view. I am talking about `string_view{}`.
Also the same with find and insert (empty string)

**Review by romange — APPROVED** [review]

**Pull request merged**